src/App.tsx

```diff
@@ -1,26 +1,177 @@
+import { Header } from './components/Header';
+import { TodoList } from './components/TodoList';
+import { Footer } from './components/Footer';
+import { ErrorNotification } from './components/ErrorNotification';
 
-const USER_ID = 0;
+export enum FilterType {
+  All = 'all',
+  Active = 'active',
+  Completed = 'completed',
+}
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [title, setTitle] = useState('');
+  const [filter, setFilter] = useState<FilterType>(FilterType.All);
+  const [errorMsg, setErrorMsg] = useState('');
+  const [loading, setLoading] = useState(false);
+  const [loadingTodoId, setloadingTodoId] = useState<number[]>([]);
+  const [isSubmiting, setIsSubmiting] = useState(false);
+
+  const visibleTodos = todos.filter(todo => {
+    switch (filter) {
+      case FilterType.Active:
+        return !todo.completed;
+      case FilterType.Completed:
+        return todo.completed;
+      case FilterType.All:
+      default:
+        return true;
+    }
+  });
+
+  const countActiveTodos = todos.filter(item => !item.completed).length;
+
+  const handleError = (message: string) => {
+    setErrorMsg(message);
+
+    setTimeout(() => {
+      setErrorMsg('');
+    }, 3000);
+  };
+
+  useEffect(() => {
+    setLoading(true);
+    service
+      .getTodos()
+      .then(setTodos)
+      .catch(error => {
+        handleError('Unable to load todos');
+        throw error;
+      })
+      .finally(() => setLoading(false));
+  }, []);
+
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (!loading && !tempTodo && !isSubmiting) {
+      inputRef.current?.focus();
+    }
+  }, [loading, tempTodo, todos.length, isSubmiting]);
+
+  function addTodo({ userId, title: newTitle, completed }: Todo) {
+    handleError('');
+    const newTempTodo: Todo = {
+      id: 0,
+      userId,
+      title: newTitle,
+      completed,
+    };
+
+    setLoading(true);
+    setTempTodo(newTempTodo);
+
+    return service
+      .addTodo({ userId, title: newTitle, completed })
+      .then(newTodo => {
+        setTodos(currentTodos => [...currentTodos, newTodo]);
+        setTitle('');
+      })
+      .catch(error => {
+        handleError('Unable to add a todo');
+        throw error;
+      })
+      .finally(() => {
+        setTempTodo(null);
+        setLoading(false);
+        inputRef.current?.focus();
+      });
+  }
+
+  function deleteTodo(todoId: number) {
+    handleError('');
+    setloadingTodoId(prev => [...prev, todoId]);
+
+    return service
+      .deleteTodo(todoId)
+      .then(() => {
+        setTodos(prev => prev.filter(todo => todo.id !== todoId));
+      })
+      .catch(() => {
+        handleError('Unable to delete a todo');
+      })
+      .finally(() => {
+        setloadingTodoId(prev => prev.filter(id => id !== todoId));
+      });
+  }
+
+  function reset() {
```

Comment:
It is important to stick to one way of creating functions, fix it everywhere
```suggestion
  const reset = () => {
```